lms/djangoapps/certificates/api.py

```diff
@@ -920,3 +922,34 @@ def _has_passed_or_is_allowlisted(course, student, course_grade):
     has_passed = course_grade and course_grade.passed
 
     return has_passed or is_allowlisted
+
+
+def invalidate_certificate_legacy_and_new(user_id, course_key_or_id):
+    """
+    Invalidate the user certificate in a given course if it exists and the user is not on the allowlist for this
+    course run.
+
+    This function is called in services.py and handlers.py within the certificates folder. As of now,
+    The call in services.py occurs when an exam attempt is rejected in the legacy exams backend, edx-proctoring.
+    The call in handlers.py is occurs when an exam attempt is rejected in the newer exams backend, edx-exams.
+    """
+    course_key = _get_key(course_key_or_id, CourseKey)
+    if _is_on_certificate_allowlist(user_id, course_key):
+        log.info(f'User {user_id} is on the allowlist for {course_key}. The certificate will not be invalidated.')
+        return False
+
+    try:
+        generated_certificate = GeneratedCertificate.objects.get(
+            user=user_id,
+            course_id=course_key
+        )
+        generated_certificate.invalidate(source='certificate_service')
```

Comment:
I think source might be a new parameter to this function. This is now called from multiple sources one is still `certificate_service` and the other is new (maybe `exam_event`?)

Reply:
Ah good point, I missed that. Will add that as a parameter to this function.